[CS-3663] Card drop banner display control

oidoug opened the pull request:
<!-- This is a pull request. Please make sure any applicable bullet points have been covered. -->

### Description

- Adds hook useWelcomeCtaBanner to control the display of the welcome banner;
- Adds rtk-query for Hub Public API.

new:
- Refactored hub-public-api to conform to new HUB structure.
- Renamed helper fetchHubBaseQuery to fetchHubAuthenticatedQuery to avoid confusion. Base only sets baseURL and base headers.

~~There are some refactoring I could do here after we merge #811 , we can move the hub public API introduced here to the hub folder and extract hub-public-api `baseQuery` function to `hub-service.ts`, so we can reuse some bits from it, but I think it's preferable to handle the public part as a different API since it doesn't need authentication.~~

- [x] Completes #(CS-3663)

### Checklist

- [x] All UI changes have been tested on a small device

### Screenshots

<img width="300" alt="Screenshot" src="https://user-images.githubusercontent.com/129619/165089089-73f20f6f-8615-4087-8492-565fae147658.jpg">

Commit:
Adds hook useWelcomeCtaBanner to control display of welcome banner;
Adds rtkquery for Hub Public API.

Commit:
Simplifies query response

danibonilha commented on cardstack/src/services/hub-public-api.ts:
```diff
@@ -0,0 +1,32 @@
+  }),
+  tagTypes: [...Object.values(HubPublicCacheTags)],
+  endpoints: builder => ({
+    getEoaClaimed: builder.query<any, EoaClaimedParams>({
```
Let's type the query return, we can also transform the response to just return the claimed boolean

Commit:
Cleans up banner hook a bit;
Adds better handled response for hub public api call.

danibonilha commented on cardstack/src/services/hub-public-api.ts:
```diff
@@ -0,0 +1,32 @@
+
+      return headers;
+    },
+  }),
```
Agreed that this can be a diff api, but we could think about customizing the `fetchHubBaseQuery` to conditionally add the token, so we can reuse the logic on choosing the hubUrl and the common headers on baseQuery

danibonilha commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -0,0 +1,56 @@
+  // Addresses that have already claimed a prepaidcard don't see the banner.
+  const hasClaimed = useMemo(() => {
+    return data?.data?.attributes?.claimed || false;
+  }, [data]);
```
By transforming the response on the api we may not even need this as the default value can be added on the query destructuring

Commit:
Refactored hub-public-api to conform to new structure.
Renamed helper fetchHubBaseQuery to fetchHubAuthenticatedQuery.

danibonilha commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -0,0 +1,54 @@
+    );
+
+    return currentWalletAddress?.index === 0 || false;
+  }, [wallets, selectedWallet, accountAddress]);
```
I believe it might be safer to rely on [web3.eth.getAccounts()](https://web3js.readthedocs.io/en/v1.2.11/web3-eth.html#getaccounts)  and check if the first address is same as the current selected wallet, instead of using the indexes locally set on wallet creation

oidoug replied:
My understanding is that we can't get addresses from the contract without loading up the HDProvider, which prompts the touch id check. I tried calling without setting the provider and it returns empty. So not sure it's ideal.
Maybe there's another way to use it I'm not aware of, plz let me know.

The index is saved when creating an address for the wallet, and its also used when creating new addresses, so it seams reasonable to use it.

danibonilha commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -0,0 +1,54 @@
+  const hasClaimed = useMemo(() => claimedResponse ?? true, [claimedResponse]);
+
+  return {
+    showBanner:
+      showBannerUserDecision && isFirstAddressForCurrentWallet && !hasClaimed,
```
we could create this const, just to cleanup and I guess we should probably check if it's layer2 to show the banner

danibonilha commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -0,0 +1,54 @@
+  }, [wallets, selectedWallet, accountAddress]);
+
+  // Addresses that have already claimed a prepaidcard don't see the banner.
+  const hasClaimed = useMemo(() => claimedResponse ?? true, [claimedResponse]);
```
nit: maybe we don't need this, we can default to true on line 17, `const { data: hasClaimed = true } = `

Commit:
Updates conditions on welcome cta banner;
Updates fetchHubBaseQuery to use extraOptions as flag for appending authentication headers.

Commit:
Adds missing memo dependency;
Fixes potential undefined exception on public api transform response

Commit:
Merges hub-public-api with hub-api for simplicity,
since the new extraOptions param for basequery solves
the auth difference.

oidoug commented:
@danibonilha, I think you'll like what I found :) by using `extraOptions` parameter for BaseQueryFn, I was able to merge the public and private endpoints, so we only have `hub-api` now.

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -16,6 +22,7 @@ export const hubApi = createApi({
   endpoints: builder => ({
     getCustodialWallet: builder.query<GetCustodialWalletQueryResult, void>({
       query: () => routes.custodialWallet,
+      extraOptions: { authenticate: true },
```
👏

danibonilha commented on cardstack/src/services/hub/hub-service.ts:
```diff
@@ -13,32 +13,42 @@ import logger from 'logger';
   unknown,
-  FetchBaseQueryError
+  FetchBaseQueryError,
+  BaseQueryExtraOptions // extraOptions
```
Maybe authenticate can default to true ? Giving 90% of the call will be authenticated

oidoug replied:
well noted, I think so, will update

Commit:
Fixes endpoint path for getEoaClaimed

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -7,11 +7,15 @@ import { fetchHubBaseQuery, hubBodyBuilder } from './hub-service';
 
 const routes = {
-  custodialWallet: '/custodial-wallet',
+  custodialWallet: '/api/custodial-wallet',
```
```suggestion
  custodialWallet: '/custodial-wallet',
```

oidoug replied:
ah, left-over from the rebase, thanks for noticing it.

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -7,11 +7,15 @@ import { fetchHubBaseQuery, hubBodyBuilder } from './hub-service';
-  custodialWallet: '/custodial-wallet',
+  custodialWallet: '/api/custodial-wallet',
   emailDrop: '/email-card-drop-requests',
+  emailCardDropRequest: ({ eoa }: EoaClaimedArg) =>
+    `/email-card-drop-requests?eoa=${eoa}`,
```
The idea of the routes object is to list all the possible routes, but here it might be confusing to  have the same route twice with the only diff being one accepts a param, for this case it might make more sense to construct the variable on the query, I'll add the suggestion.

danibonilha commented on cardstack/src/services/hub/hub-api.ts:
```diff
@@ -31,16 +36,25 @@ export const hubApi = createApi({
       }),
     }),
+    getEoaClaimed: builder.query<EoaClaimedResultType, EoaClaimedArg>({
+      query: routes.emailCardDropRequest,
```
```suggestion
      query: ({ eoa }) => `${routes.emailDrop}?eoa=${eoa}`,
```

danibonilha commented on cardstack/src/services/hub/hub-types.ts:
```diff
@@ -7,3 +7,14 @@ export type GetCustodialWalletQueryResult = KebabToCamelCaseKeys<CustodialWallet
+  authenticate?: boolean;
+}
+
+export interface EoaClaimedArg {
+  eoa: string;
+}
```
nit: It would be nice to follow the pattern as the other types something like `GetEoaClaimedQueryParams`

danibonilha reviewed (approved):
Props on adding the extraParams 👏 , it's looking good, some minor nits, and one required change -> remove the  /api/ leftover on custodialWallet of hub routes.

Commit:
Fixes custodialWallet route;
Updates naming for GetEoa types;
Defaults extraOptions authenticate to true.

Commit:
Skipping GetEoaClaimedQuery call while address not ready.
